File: prime_dev_6.rb
def div_while_possible(result, n, prime)
	quot, rem = n.divmod(prime)
	while rem == 0 do
		result.push prime
		quot, rem = quot.divmod(prime)
	end
	return quot * prime + rem
end

n = ARGV.shift.to_i
result = []

n = div_while_possible(result, n, 2)
n = div_while_possible(result, n, 3)

m = 1 # 6m-1, 6m+1のみ考えれば十分
limit = Math::sqrt(n).to_i # nの平方根より大きい素因数は存在しない

while n > limit do
	n = div_while_possible(result, n, 6 * m - 1)
	n = div_while_possible(result, n, 6 * m + 1)

	m = m + 1
end

p result.to_s

File: primeDev6.go
package main

import(
	"fmt"
	"math"
)

func divWhilePossible(n, prime int) (newN int, result []int) {	
	var rem int

	for {
		rem = n % prime
		n = n / prime
		
		if rem != 0 {
			break
		} else {
			result = append(result, prime)
		}
	}
	
	newN = n * prime + rem
	return
}

func main() {
	m := 1
	n := 12
	// var result, newResult []int
	var result, newResult []int

	n, newResult = divWhilePossible(n, 2)
	result = append(result, newResult...)

	n, newResult = divWhilePossible(n, 3)
	result = append(result, newResult...)
	
	limit := int(math.Sqrt(float64(n)))
	for n > limit {
		n, newResult = divWhilePossible(n, 6 * m - 1)
		result = append(result, newResult...)

		n, newResult = divWhilePossible(n, 6 * m + 1)
		result = append(result, newResult...)
		
		m++
	} 
	
	fmt.Println(result)
}